src/commons/editor/useComments.tsx

```diff
@@ -0,0 +1,261 @@
+import { require as acequire, Ace } from 'ace-builds';
+import * as React from 'react';
+import ReactDOM from 'react-dom';
+
+import { groupBy, map, omit, sortBy, values, keyBy } from 'lodash';
+
+import { EditorHook } from './Editor';
+import Comments, { IComment } from './Comments';
+import { v1 as uuidv1 } from 'uuid';
+import { EventEmitter } from 'events';
+const LineWidgets = acequire('ace/line_widgets').LineWidgets;
+
+// Inferred from: https://github.com/ajaxorg/ace/blob/master/lib/ace/ext/error_marker.js#L129
+interface IWidget {
+  row: number;
+  fixedWidth: boolean;
+  coverGutter: boolean;
+  el: Element;
+  type: string;
+}
+
+interface ILineManager {
+  attach: (editor: Ace.Editor) => void;
+  addLineWidget: (widget: IWidget) => void;
+  removeLineWidget: (widget: IWidget) => void;
+}
+
+export interface CommentAPI {
+  // Assumes commentId is immutable. Do not update it unless explicitly via API.
+  updateComment: (...comments: IComment[]) => void;
+  updateCommentsBeingEdited: (...comments: IComment[]) => void;
+  removeComment: (comment: IComment) => void;
+  removeCommentEdit: (comment: IComment) => void;
+  // Probably need a replaceId for comments to update when the server does.
+
+  isUnsubmittedComment: (comment: IComment) => boolean;
+}
+
+/*  NOTE ON INTEGRATION WITH SERVER SIDE:
+
+All the data is bound in this component so that it can easily be updated from server side.
```

Comment:
Whilst this seems ideal, I predict a few problems already with managing the data flow through the saga (backend middleware), Redux store, workspace and the editor.

Reply:
I'm actually thinking we leave out redux, load assessment id from props. Then sync via real time pub sub only.

Reply:
Hmm, but that would still require to Editor to recognise which assessment and question it's displaying - not too sure if we can entirely avoid using data from the Redux store.